### ISSUE TITLE
False positive triggered by a combination of dataclasses, inheritance and circular references

### ISSUE THREAD

**Issue by zivanfi** [microsoft/pyright#523]
**Describe the bug**

Pyright complaints that a parameter does not exist, even though it does.

**To Reproduce**

Use pyright on the code pasted below.

**Expected behavior**

The parameter should be accepted.

**Screenshots or Code**

Example code to reproduce the error:

```
from dataclasses import dataclass
from typing import List
from weakref import ref, ReferenceType  # pylint: disable=unused-import


@dataclass
class BaseClass:
    my_ref: "ReferenceType[ReferredClass]"


@dataclass
class SubClass(BaseClass):
    pass


@dataclass
class ReferredClass:
    sub_class: List[SubClass]

    def trigger_bug(self):
        SubClass(my_ref=ref(self))
```

Output of running pyright:

```
/tmp $ pyright a.py 
typingsPath /tmp/typings is not a valid directory.
Searching for source files
Found 1 source file
/tmp/a.py
  21:18 - error: No parameter named 'my_ref'
1 error, 0 warnings 
Completed in 1.328sec
```

**VS Code extension or command-line**

Initially both VS Code extension and command-line usage of pyright 1.1.22 were affected. As I was trying to produce a minimal example, vscode seemed to behave strangely (gave different results to the same code after making a change and then undoing it). I continued reducing the code using the command line tool, which seemed to be more consistent. In its current form the example does not seem to trigger an error in vscode, only from the command line.

**Additional context**

Moving `my_ref` from `BaseClass` to `SubClass` makes the error disappear. Removing `sub_class` from `ReferredClass` makes the error disappear.

Using a normal reference instead of a weak one does not change the result.

In case the question arises whether this code makes sense in spite of the circular reference: ReferredClass can be created with an empty collection and then that collection can be populated  with instances of SubClass that each contain a reference to ReferredClass.

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the problem thanks to your clear description.

The problem is related to the circular dependency. When a normal class is referenced, Pyright doesn't analyze the types of its members until they are referenced. In the case of a dataclass, Pyright needs to analyze members and their types immediately to synthesize a constructor (`__init__`) method with the member names and types. Because of the circular reference in your example, there is an order dependency. That explains why the error goes away when you start interacting with the code.

Now that I understand the problem, I'll need to think more about how to fix it.

**Comment by erictraut** [comment]
I found a way to handle the circular type references. This will be fixed in the next release of Pyright.

**Comment by zivanfi** [comment]
Great, thanks!

**Comment by erictraut** [comment]
This is included in 1.1.23, which I just published.
